sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java

```diff
@@ -159,4 +180,70 @@ ShardRecordsIterator createShardIterator(SimplifiedKinesisClient kinesis,
     return new ShardRecordsIterator(checkpoint, kinesis);
   }
 
+  /**
+   * Waits until all records read from given shardRecordsIterator are taken from
+   * {@link #recordsQueue} and acked.
+   * Uses {@link #numberOfRecordsInAQueueByShard} map to track the amount of remaining events.
+   */
+  private void waitUntilAllShardRecordsRead(ShardRecordsIterator shardRecordsIterator)
+      throws InterruptedException {
+    while (!allShardRecordsRead(shardRecordsIterator)) {
+      Thread.sleep(TimeUnit.SECONDS.toMillis(1));
+    }
+  }
+
+  private boolean allShardRecordsRead(final ShardRecordsIterator shardRecordsIterator) {
+    return numberOfRecordsInAQueueByShard.get(shardRecordsIterator.getShardId()).get() == 0;
```

Comment:
Does the value for a given shard id obey a strict (increase -> decrease to 0) pattern, or can it get back to non-0 after getting to 0? If yes, this can be racy.

Reply:
The counter goes up and down during normal operation when shard is in open state. But here we are waiting until all records from already closed shard are processed so there is no way this counter can ever go up for this shard.

Reply:
Consider inlining this method, and adding a comment that, by the time we get here, no more records can be added to this shard because (explanation) so it's strictly decreasing?